### PR TITLE
Refactor PianoRoll

### PR DESCRIPTION
**Pull request by Veratil**
Kind of a bigger PR, but fixes a few things: #4098, #5008 (partial, but easy to add to editors now).

In addition to those fixes, it reworks how the piano roll was painted before (and makes it somewhat easier to follow). Local tests of ~1000 `paintEvent` frames using the demo `Skiessi-C64.mmpz` put `master`'s `paintEvent` at ~6-6.1ms average time, this PR runs at ~4-4.1ms (improvement of ~33%) on `DEBUG` build.

General summary of updates:
* Split out SongEditor's `positionLine` class into an updated separate class file `PositionLine`
* No longer can the key area be resized larger than `NumKeys * KEY_LINE_HEIGHT` (108 * 12) pixels
* If window is sized larger than max key area size, note edit area is resized instead
* Key area snaps to `KEY_LINE_HEIGHT` to prevent partial key draws (except if top key is black key, the previous key is drawn in that case)
* Position marker added to timeline
* Simplifies some math, some if branches, and draw code (old code went through white keys then through black keys, now only one for loop)

EDIT: I've updated the title to be a little more general since I'm touching a little more than just `painteEvent` with this PR. It still focuses on `paintEvent`, but there's a lot more to it at this point.

### CODE REVIEW

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp]: It seems like you've added vertical lines to the note edit area. However, the color is inconsistent with that of the key area. Can't you extend the vertical line for the key area instead if you're going to keep those lines drawn?

**Review by PhysSong — COMMENTED** [review]
Here's my initial review for this PR.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp]: Yes... yes it would be. :joy:

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp]: I figured this is "intended" behavior, but we can always revert it if necessary.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp]: Future enhancement. I'm currently reworking the mouse events. Once that is done I'm going to work on moving things to where they should be. I'd rather rework first to make sure behavior stays the same, then start moving things to where they should be handled.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp]: Ah! I see what you mean. Minor oversight. Will fix that up. 👍

**Comment by JohannesLorenz** [comment]
@Veratil This is marked WIP, but I think it's waiting for code review? Do you plan more enhancements/fixes in this PR currently?

**Comment by Veratil** [comment]
> 
> 
> @Veratil This is marked WIP, but I think it's waiting for code review? Do you plan more enhancements/fixes in this PR currently?

Was waiting for other review/testing. It just needs to be rebased with recent updates it seems.

**Comment by JohannesLorenz** [comment]
OK, removed "in progress".

Whoever reads this, please consider doing a review or testing the PR.

**Comment by JohannesLorenz** [comment]
I'll do the review now (as we concluded that a review is still missing).

File: src/gui/editors/PianoRoll.cpp
Comment (JohannesLorenz) [reply src/gui/editors/PianoRoll.cpp]: Note: This will not be solved in this PR. I'll still leave it open as an information.

**Comment by JohannesLorenz** [comment]
There's a problem with the PR base. You say you solve #4098, which is a 1.2.2 issue, but you want to merge into master, which will not fix that issue in 1.2.2.

So you need to decide:

1. Either rebase the whole PR on stable-1.2, or
2. Leave it on master and do not claim to solve 4098, or
3. Split it and only base the 4098 fix on stable-1.2 

I'd say use 3 if possible (It's a large amount of changes for a feature-freeze branch currently), if not use 1.

File: src/gui/editors/PianoRoll.cpp
Comment (JohannesLorenz) [inline src/gui/editors/PianoRoll.cpp]: This looks right with 4-space-tabs, but in case someone uses invalid tabs, I'd rather change this:

```suggestion
		    -4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
```

**Review by JohannesLorenz — DISMISSED** [review]
Style/Code/Test Review finished, comments added.

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [reply src/gui/editors/PianoRoll.cpp]: It tests if the key area is tall enough to draw all keys.

Comment (JohannesLorenz) [reply src/gui/editors/PianoRoll.cpp]: OK, now I think it's clear, indeed. I'll resolve this comment.

**Comment by Veratil** [comment]
> 
> 
> There's a problem with the PR base. You say you solve #4098, which is a 1.2.2 issue, but you want to merge into master, which will not fix that issue in 1.2.2.
> 
> So you need to decide:
> 
>     1. Either rebase the whole PR on stable-1.2, or
> 
>     2. Leave it on master and do not claim to solve 4098, or
> 
>     3. Split it and only base the 4098 fix on stable-1.2
> 
> 
> I'd say use 3 if possible (It's a large amount of changes for a feature-freeze branch currently), if not use 1.

Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.

**Comment by JohannesLorenz** [comment]
> Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.

This issue is a crash on stable-1.2 and thus *must* be fixed there in some way (which excludes 2 btw). If you do (4), someone needs to cherry pick it to stable-1.2, and I don't know anyone who could cherry-pick it better then the author...

**Comment by Veratil** [comment]
> 
> 
> > Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.
> 
> This issue is a crash on stable-1.2 and thus _must_ be fixed there in some way (which excludes 2 btw). If you do (4), someone needs to cherry pick it to stable-1.2, and I don't know anyone who could cherry-pick it better then the author...

That's understandable. There's two ways I can see to fix `stable-1.2`: 1) set the max height of the window (headers + pianoroll max + note edit min); or 2) stop drawing keys like this PR does.

For the max height idea, I was attempting to figure out how to set the max height first (but as mentioned this restricts the note edit window height at some point), but my efforts led to just the "mainWindow" of the piano roll being maxed. The actual window could be resized while the height of the `PianoRollWindow` would stop. It was very weird to me. I decided to work on rewriting the paintEvent instead. :)

As mentioned before, I would still say this PR fixes the issue though since iirc it happened on master as well hence me looking into this entire thing, so it is acceptable to mark it as such IMO.

**Comment by JohannesLorenz** [comment]
> As mentioned before, I would still say this PR fixes the issue though since iirc it happened on master as well hence me looking into this entire thing, so it is acceptable to mark it as such IMO.

If you want to do this I'll create a duplicate of the issue for stable-1.2, which can then be fixed by a e.g. a cherry-picked PR.

**Comment by JohannesLorenz** [comment]
@Veratil Do you plan to finish this PR? It's almost done...

File: include/PianoRoll.h
Comment (Veratil) [reply include/PianoRoll.h]: The latest commit tries to use forward declaration, seems it failed to build though...

Comment (JohannesLorenz) [reply include/PianoRoll.h]: It should have failed on your local computer, too.
You can fix it by adding the include into the affected cpp files.

Comment (Veratil) [reply include/PianoRoll.h]: I think it succeeded since I did an incremental build initially and it didn't complain. It fails now. 🤷‍♂️

**Comment by JohannesLorenz** [comment]
Code and style are OK.

What remains is that someone needs to check whether it fixes #4098 - My monitor is not large enough.

**Comment by PhysSong** [comment]
@JohannesLorenz You can move down the piano-roll window and resize it from its upper side. When you repeat that enough times, you'll get a large enough window.

**Comment by JohannesLorenz** [comment]
Thanks @PhysSong that made it.

There's one issue still, though I'm not sure if we need to fix it: Open a piano roll, resize piano roll so it's getting big. Move it down to the bottom and resize it to the upper limit (that way, you enlarge the PianoRoll, thanks to @PhysSong ). Do that 3 or 4 times. Play a few notes. The lower notes are sometimes green covered, and the PianoRoll marks the wrong keys when I play keys. (It's not the same bug that I report for the stable PR)

**Comment by JohannesLorenz** [comment]
I'm 100% happy with the PR. Testing finished.

Can I squash-and-merge it?

**Comment by PhysSong** [comment]
> Can I squash-and-merge it?

@JohannesLorenz I think not now. #5442 may affect this PR.

**Comment by Veratil** [comment]
> 
> 
> > Can I squash-and-merge it?
> 
> @JohannesLorenz I think not now. #5442 may affect this PR.

We need to decide which order to merge things, that's all.

**Comment by PhysSong** [comment]
@Veratil Now it's your turn to fix conflicts with #5442. :smile:

**Comment by Veratil** [comment]
> 
> 
> @Veratil Now it's your turn to fix conflicts with #5442. 😄

Rebased and adjusted code. 👍 Did some basic tests and zooms work still. Need a tester or two now!

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:2951]: You can't use these constants anymore because of #5442.

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:2976]: The compiler warns they may be used uninitialized, while it's not. Those warnings will make CI builds to fail.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:2951]: I switched to using the one from the PR, and all the labels are way off now. 😩

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:2976]: Will be fixed in next commit.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:2951]: I've managed to fix the offsets by doing a little different calculations, but with Y zoom I've found there are some draw bugs now when certain keys are the top key.

File: data/themes/default/style.css
Comment (Veratil) [inline data/themes/default/style.css:168]: I would like to note these updates are purely for testing and by no means are to be considered final.

**Comment by PhysSong** [comment]
I found a new bug.
1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
2. Change the vertical zoom level to 25%
3. The key area will disappear, and the resize bar is at a wrong place

**Comment by Veratil** [comment]
> 
> 
> I found a new bug.
> 
>     1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
> 
>     2. Change the vertical zoom level to 25%
> 
>     3. The key area will disappear, and the resize bar is at a wrong place

Interesting. The resize bar is in the correct place actually, but the keys didn't get drawn for some reason. If you drag the resize bar up one pixel everything appears properly. I'll look into why they don't get drawn though.

**Comment by Veratil** [comment]
> 
> 
> > I found a new bug.
> > ```
> > 1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
> > 
> > 2. Change the vertical zoom level to 25%
> > 
> > 3. The key area will disappear, and the resize bar is at a wrong place
> > ```
> 
> Interesting. The resize bar is in the correct place actually, but the keys didn't get drawn for some reason. If you drag the resize bar up one pixel everything appears properly. I'll look into why they don't get drawn though.

Added some `printf`'s to see what was causing the issue. It looks like `m_startKey` is getting set to `-108`. Added some bounds checking to `m_startKey` and `m_totalKeysToScroll` so they would always be minimum 0.

**Comment by PhysSong** [comment]
This one conflicts with #5543. @Veratil Could you resolve conflicts so we can test this one more time?

**Comment by Veratil** [comment]
> 
> 
> This one conflicts with #5543. @Veratil Could you resolve conflicts so we can test this one more time?

Updated!

**Comment by PhysSong** [comment]
If I select `classic` theme, I can't see keys.

**Comment by Veratil** [comment]
> 
> 
> If I select `classic` theme, I can't see keys.

Oops, that's because I forgot to add the same properties to `classic`.

**Comment by Spekular** [comment]
Just want to note that after this is merged, #669 should be easier to implement.

**Comment by Veratil** [comment]
@Spekular @PhysSong do we feel this is safe to merge yet? I want to do more, but I don't want to add to the huge changes for this already.

**Comment by Spekular** [comment]
I haven't done a review so I'm not sure my opinion bears much weight (nothing stood out as problematic on a quick skim FWIW). What I can say is that I'm not opposed to the goals of the PR as described in the OP, all the changes mentioned sound good. Aside from that I'd like to defer to Johannes and PhysSong's judgement :P

**Comment by JohannesLorenz** [comment]
I haven't watched it since April, so I can't say whether it's OK.

For the question about using a new PR, I'm always for making multiple small PRs - as you say, this one already got large enough. :smile:

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:2685]: Very minor, but `pianoAreaHeight` looks clearer.

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:3302]: The same.

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:2890]: If `lastKey + 1` is a white key, the loop will stop after drawing it, whereas it will draw `lastKey` if `lastKey + 1` is a black key. Is this the intended behavior?

**Comment by PhysSong** [comment]
Also, you changed the width of black keys to the same as white keys. Although [it's themeable](https://github.com/LMMS/lmms/issues/3751#issuecomment-623147961), don't we need discussions when changing the default?

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:2890]: This handles the case where we draw a black key over a white key (current behavior). So, yes it's intended for that purpose.

**Comment by Veratil** [comment]
> 
> 
> Also, you changed the width of black keys to the same as white keys. Although [it's themeable](https://github.com/LMMS/lmms/issues/3751#issuecomment-623147961), don't we need discussions when changing the default?

Accidental leftover while testing. Changed down to 48 pixels.

**Comment by PhysSong** [comment]
There might be something I missed, but I think it looks okay now.

**Comment by Veratil** [comment]
If no one opposes, I'll merge this in 2 days.

**Pull request merged**
